### PR TITLE
chore(magmad): magmad is bazelified

### LINKED ISSUE #11284: bazel: Bazelify Python magmad

### PR DESCRIPTION
**Pull request by LKreutzer**
Signed-off-by: Lars Kreutzer <lars.kreutzer@tngtech.com>

## Summary

- Module magmad is bazelified including the tests.
- Please note: 
  - Suggestion: Review commit by commit - first the submodules, then swagger support, magmad service, magmad tests.
  - See documentation in swagger commit for origin of approach. This approach can easily be extended for other languages (Go, C, C++)
  - In analogy to the make files aioh2 is installed via git repository and not via pip.
  - Swagger produces some logging (if not cached) ~20 lines. Is this acceptable? Or should this be turned off or even configurable?

## Test Plan

-  Run magmad: 
   - `bazel run //orc8r/gateway/python/magma/magmad:magmad`
   - Note: orc8r connection is needed in order to run this service in magma-vm.
- Run test: 
  - `bazel test //orc8r/gateway/python/magma/magmad/...:all`

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Comment by LKreutzer** [comment]
Force pushed aioh2 repository configuration (to fix bazel warning).

File: orc8r/gateway/python/magma/magmad/check/kernel_check/tests/BUILD.bazel [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+load("//bazel:python_test.bzl", "pytest_test")
+
+COMMON_TEST_DEPS = ["//orc8r/gateway/python/magma/magmad/check/kernel_check:kernel_versions"]
```
Comment (themarwhal) [inline orc8r/gateway/python/magma/magmad/check/kernel_check/tests/BUILD.bazel]: It might be cleaner to just place the dependency in the pytest_test macro? (But I'm open to leaving it like this if you have reasons)

File: bazel/python_swagger.bzl [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+#     Copyright (c) 2021 Meetup, Inc.
+#     https://github.com/meetup/rules_openapi/blob/86fa11d0a8a8188ceecb74b5674af3f7363701e8/openapi/openapi.bzl
+# We thank the authors of [1] and [2].
+
+""" Provides support for building python code from swagger specifications. """
+
+load("@bazel_tools//tools/build_defs/repo:jvm.bzl", "jvm_maven_import_external")
+
+SWAGGER_CLI_VERSION = "2.4.16"
+SWAGGER_CLI_SHA256 = "154b5a37254a3021a8cb669a1c57af78b45bb97e89e0425e3f055b1c79f74a93"
```
Comment (themarwhal) [inline bazel/python_swagger.bzl]: This might be mentioned somewhere but where do you get the sha value from?

File: orc8r/gateway/python/magma/magmad/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+py_binary(
+    name = "magmad",
+    srcs = SRCS,
+    imports = [ORC8R_ROOT],
+    # legacy_creat_init = False is required to fix issues in module import, see https://github.com/rules-proto-grpc/rules_proto_grpc/issues/145
+    legacy_create_init = False,
+    main = "main.py",
+    python_version = "PY3",
+    visibility = ["//visibility:public"],
+    deps = DEPS,
```
Comment (themarwhal) [inline orc8r/gateway/python/magma/magmad/BUILD.bazel]: A general comment with no specific solution in mind.. 

This large list of dependency worries me that it'll be harder for people to keep these up to date. I'm thinking of cases where we remove dependencies from the code but we forget to remove dependencies from the build file. I was wondering if creating smaller `py_library`s would help, but I am not sure. What is your opnion here on helping devs manage these in the long run?

File: .bazelrc [hunk 4]
```diff
@@ -28,6 +28,11 @@ build:devcontainer --disk_cache=/workspaces/magma/.bazel-cache
 common:devcontainer --repository_cache=/workspaces/magma/.bazel-cache-repo
 build:devcontainer --define=folly_so=1
 
+# TEST CONFIGS
+#  Bazel test runtime default: PATH=/bin:/usr/bin:/usr/local/bin
+#  Some python tests require access to /usr/sbin binaries (e.g. route, ifconfig)
+build --test_env=PATH=/bin:/usr/bin:/usr/local/bin:/usr/sbin
```
Comment (Neudrino) [inline .bazelrc:34]: I usually know the form `PATH=/newly/added:$PATH` from bash, which makes it much clearer that only one part is added anew.
Would that (or something similar) be possible here?

File: bazel/python_swagger.bzl (hunk 2 above)
Comment (LKreutzer) [reply bazel/python_swagger.bzl]: This was copied from the sources mentioned in the file. But we verified it: 
`$ sha256sum swagger-codegen-cli-2.4.16.jar 
154b5a37254a3021a8cb669a1c57af78b45bb97e89e0425e3f055b1c79f74a93  swagger-codegen-cli-2.4.16.jar`
Too bad, maven central does not provide sha256. Easy way via bazel (but security net is kind of circumvented): remove argument "artifact_sha256", do bazel build, find output 
`INFO: SHA256 (https://repo.maven.apache.org/maven2/io/swagger/swagger-codegen-cli/2.4.16/swagger-codegen-cli-2.4.16.jar) = 154b5a37254a3021a8cb669a1c57af78b45bb97e89e0425e3f055b1c79f74a93`

File: orc8r/gateway/python/magma/magmad/check/kernel_check/tests/BUILD.bazel (hunk 1 above)
Comment (LKreutzer) [reply orc8r/gateway/python/magma/magmad/check/kernel_check/tests/BUILD.bazel]: Yes, you are right - will be changed!

File: orc8r/gateway/python/magma/magmad/BUILD.bazel (hunk 3 above)
Comment (LKreutzer) [reply orc8r/gateway/python/magma/magmad/BUILD.bazel]: Good point, we were already thinking about splitting main.py from the remaining source files (e.g. magmad_lib), however further splitting up requires a high-level understanding of the magmad functionality. We will try the magmad_lib approach and see whether this simplifies things.

File: bazel/python_swagger.bzl [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+def _generator_command(ctx, gen_dir):
+    java_path = ctx.attr._jdk[java_common.JavaRuntimeInfo].java_executable_exec_path
+
+    gen_cmd = str(java_path)
+    gen_cmd += " -cp {cli_jar}".format(cli_jar = ctx.file._swagger_cli.path)
+    gen_cmd += " io.swagger.codegen.SwaggerCodegen generate -i {spec} -l python -o {output}".format(
+        spec = ctx.file.spec.path,
+        output = gen_dir,
+    )
+    gen_cmd += " -Dmodels "
```
Comment (Neudrino) [inline bazel/python_swagger.bzl]: We usually only prepend whites-paces to all commands here. Is there a special reason there needs also to be one postpended?

File: bazel/python_swagger.bzl [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    component = ctx.file.spec.path.split("/")[0]
+    gen_dir = "{}/{}/swagger".format(declared_dir.path, component)
+
+    commands = [
+        "mkdir -p {}".format(gen_dir),
+        _generator_command(ctx, gen_dir),
+        # Build required folder structure and remove unused build outputs.
+        "rm -rf {}/test".format(gen_dir),
+        "mv {}/swagger_client/models {}".format(gen_dir, gen_dir),
+        "rm -rf {}/swagger_client".format(gen_dir),
```
Comment (Neudrino) [inline bazel/python_swagger.bzl:71]: _No action_: No I see, why you mentioned that swagger is somewhat "hacky"...

Comment (Neudrino) [inline bazel/python_swagger.bzl:71]:
```suggestion
        "mv {}/swagger_client/models {}".format(gen_dir, gen_dir),
        "rm -rf {}/{swagger_client,test}".format(gen_dir),
```
Might that be a simplification?

**Review by Neudrino — APPROVED** [review]
LGTM overall.

File: .bazelrc (hunk 4 above)
Comment (LKreutzer) [reply .bazelrc:34]: No this is not possible here since $PATH would reference the PATH variable from the host system - and not the temporary PATH variable from the bazel runtime.

File: bazel/python_swagger.bzl (hunk 5 above)
Comment (LKreutzer) [reply bazel/python_swagger.bzl]: Okay, done.

File: bazel/python_swagger.bzl (hunk 6 above)
Comment (LKreutzer) [reply bazel/python_swagger.bzl:71]: The current approach is imho more readable.

File: orc8r/gateway/python/magma/magmad/BUILD.bazel (hunk 3 above)
Comment (LKreutzer) [reply orc8r/gateway/python/magma/magmad/BUILD.bazel]: Force pushed the main.py/library approach, the dependencies are now split somewhat more. What is your opinion on this? We suggest to use this approach as a coding pattern for future services.

File: orc8r/gateway/python/magma/magmad/BUILD.bazel [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    "//orc8r/gateway/python/magma/magmad/check/network_check:routing_table",
+    "//orc8r/protos:metricsd_python_grpc",
+    "//orc8r/gateway/python/magma/common:stateless_agw",
+    "//orc8r/protos:magmad_python_grpc",
+    "//orc8r/gateway/python/magma/common:cert_validity",
+    "//orc8r/protos:state_python_grpc",
+    "//orc8r/protos:sync_rpc_service_python_grpc",
+    "@aioh2_repo//:aioh2",
+    "//orc8r/gateway/python/magma/common:streamer",
+]
```
Comment (Neudrino) [inline orc8r/gateway/python/magma/magmad/BUILD.bazel]: All other lists are sorted lexicographical. Would we also want to do this for this list?

**Comment by nstng** [comment]
New dependency in experimental/bazel-base needed. Fix ongoing - will be added as additional commit here.

File: bazel/python_swagger.bzl (hunk 2 above)
Comment (themarwhal) [reply bazel/python_swagger.bzl]: I think I will forget this information ^ at some point. Can we add it in as a comment? :)

File: orc8r/gateway/python/magma/magmad/BUILD.bazel (hunk 7 above)
Comment (themarwhal) [reply orc8r/gateway/python/magma/magmad/BUILD.bazel]: `bazel run //:buildifier` should take care of it :)

File: orc8r/gateway/python/magma/magmad/check/network_check/BUILD.bazel [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+load("@rules_python//python:defs.bzl", "py_library")
+
+ORC8R_ROOT = "../../../../"
+
+SUBPROCESS_WORKFLOW_DEP = "//orc8r/gateway/python/magma/magmad/check:subprocess_workflow"
```
Comment (themarwhal) [inline orc8r/gateway/python/magma/magmad/check/network_check/BUILD.bazel]: I'd also rather have just the dependency specified directly, I think. Just so I don't have to jump around in the file to read.

File: orc8r/gateway/python/magma/magmad/generic_command/tests/BUILD.bazel [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+load("//bazel:python_test.bzl", "pytest_test")
+
+COMMON_TEST_DEPS = ["//orc8r/gateway/python/magma/magmad/generic_command:command_executor"]
```
Comment (themarwhal) [inline orc8r/gateway/python/magma/magmad/generic_command/tests/BUILD.bazel]: this too, let's just add the dep directly.

File: orc8r/gateway/python/magma/magmad/tests/BUILD.bazel [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+load("//bazel:python_test.bzl", "pytest_test")
+
+COMMON_TEST_DEPS = ["//orc8r/gateway/python/magma/magmad:magmad_lib"]
```
Comment (themarwhal) [inline orc8r/gateway/python/magma/magmad/tests/BUILD.bazel]: this one too

File: orc8r/gateway/python/magma/magmad/BUILD.bazel (hunk 7 above)
Comment (nstng) [reply orc8r/gateway/python/magma/magmad/BUILD.bazel]: Sadly, no - we will sort it manually. Are you sure it should take care of it? Maybe because we moved it into variables.

Comment (themarwhal) [reply orc8r/gateway/python/magma/magmad/BUILD.bazel]: :'(

Comment (themarwhal) [reply orc8r/gateway/python/magma/magmad/BUILD.bazel]: ohhh maybe that is why....

Comment (themarwhal) [reply orc8r/gateway/python/magma/magmad/BUILD.bazel]: In that case, do you think it's worth embedding some of these deps/source files directly?

Comment (themarwhal) [reply orc8r/gateway/python/magma/magmad/BUILD.bazel]: I would opt for placing these items directly into the macro to avoid having to jump around when reading, but open to other suggestions.

Comment (nstng) [reply orc8r/gateway/python/magma/magmad/BUILD.bazel]: I do not have a strong opinion about that. We will change this as recommended. For existing services we will keep this in mind and change it when we touch respective code.

Comment (nstng) [reply orc8r/gateway/python/magma/magmad/BUILD.bazel]: (btw tested: the buildifier works when the variables are inlined)

File: bazel/python_swagger.bzl [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+        artifact = "io.swagger:swagger-codegen-cli:" + SWAGGER_CLI_VERSION,
+        artifact_sha256 = SWAGGER_CLI_SHA256,
+        server_urls = ["https://repo.maven.apache.org/maven2"],
+        licenses = ["notice"],  # Apache 2.0 License
+    )
+
+    native.bind(
+        name = "bazel_rules_swagger/swagger-cli",
+        actual = "@maven_swagger_codegen_cli//jar",
+    )
```
Comment (themarwhal) [inline bazel/python_swagger.bzl]: Curious, why is this needed? Is there an assumption somewhere about the name?

File: bazel/python_swagger.bzl (hunk 2 above)
Comment (LKreutzer) [reply bazel/python_swagger.bzl]: Added comment.

**Review by themarwhal — APPROVED** [review]

File: orc8r/gateway/python/magma/magmad/BUILD.bazel (hunk 3 above)
Comment (themarwhal) [reply orc8r/gateway/python/magma/magmad/BUILD.bazel]: Looks great :) Thanks!

File: orc8r/gateway/python/magma/magmad/BUILD.bazel (hunk 7 above)
Comment (LKreutzer) [reply orc8r/gateway/python/magma/magmad/BUILD.bazel]: Change pushed.

File: orc8r/gateway/python/magma/magmad/check/network_check/BUILD.bazel (hunk 8 above)
Comment (LKreutzer) [reply orc8r/gateway/python/magma/magmad/check/network_check/BUILD.bazel]: Change pushed.

File: orc8r/gateway/python/magma/magmad/generic_command/tests/BUILD.bazel (hunk 9 above)
Comment (LKreutzer) [reply orc8r/gateway/python/magma/magmad/generic_command/tests/BUILD.bazel]: Change pushed.

File: orc8r/gateway/python/magma/magmad/tests/BUILD.bazel (hunk 10 above)
Comment (LKreutzer) [reply orc8r/gateway/python/magma/magmad/tests/BUILD.bazel]: Change pushed.

**Review by Neudrino — APPROVED** [review]

File: bazel/python_swagger.bzl (hunk 11 above)
Comment (LKreutzer) [reply bazel/python_swagger.bzl]: Thanks for pointing this out! Initially we thought this was needed for referencing the cli-jar in a label. After a closer look at the documentation we learned that this is actually not necessary. We removed this bit, code is cleaner now - changes pushed.

**Review by nstng — APPROVED** [review]
done in pairing

**Review by electronjoe — APPROVED** [review]

**Comment by LKreutzer** [comment]
Rebased on current master.
